### PR TITLE
12243 - Startup Global Key Commands can also send a Global Hotkey

### PR DESCRIPTION
**Pull request by Cattlesquat**

### CODE REVIEW

**Comment by riverwanderer** [comment]
#11998 relates to this, suggesting instead adapting the Action Button to achieve Startup GHKs, and proposes other features similar to normal buttons.

**Comment by Cattlesquat** [comment]
@BrentEaston Asked your review here mainly because I want to make sure the "sending the global hotkey" part is done "the right way"

**Review by BrentEaston — COMMENTED** [review]
Happy with the way you are handling the execution of the Global Hotkey.

However, I see one issue.

In the configure, the Global HotKey appears before the Global Key Command. However, in apply(), you call super.apply() to run the GKC BEFORE the GHK runs. Either they need to be reversed, or possibly a better solution is to have a drop-down to select whether you want a GHK or a GKC? Having both is a bit confusing since all the match stuff does not apply to the GHK and I would think you would usually want one or the other. If you need both, you can always add a second one of the other type?

**Comment by Cattlesquat** [comment]
@BrentEaston GREAT idea, I was worried about the confusion myself -- have changed it into a binary (you either send a hotkey or you send a global key command)

File: vassal-app/src/main/java/VASSAL/configure/VisibilityCondition.java [hunk 1]
```diff
@@ -20,4 +20,19 @@
...
+    public VisibilityAND(VisibilityCondition a1, VisibilityCondition a2) {
+      and1 = a1;
+      and2 = a2;
+    }
+
+    @Override
+    public boolean shouldBeVisible() {
+      return ((and1 == null) || and1.shouldBeVisible()) && ((and2 == null) || and2.shouldBeVisible());
+    }
+  }
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/configure/VisibilityCondition.java:37]: `VisibilityAND` should probably be in its own file, and have `final` members.

File: vassal-app/src/main/java/VASSAL/build/module/StartupGlobalKeyCommand.java [hunk 2]
```diff
@@ -139,16 +169,37 @@ public VisibilityCondition getAttributeVisibility(String key) {
...
+      //};
+      // THINGS THAT DON'T WORK
+      //return () -> (SEND_KEY_COMMAND.equals(hotkeyOrKeyCommand) && (super.getAttributeVisibility(key)).shouldBeVisible());
+      return new VisibilityCondition.VisibilityAND(() -> SEND_KEY_COMMAND.equals(hotkeyOrKeyCommand), super.getAttributeVisibility(key));
+
+      // THINGS THAT DO "WORK"
+      //return super.getAttributeVisibility(key);
+      //return () -> SEND_KEY_COMMAND.equals(hotkeyOrKeyCommand);
+      //return new VisibilityCondition.VisibilityAND(() -> SEND_KEY_COMMAND.equals(hotkeyOrKeyCommand), () -> SEND_KEY_COMMAND.equals(hotkeyOrKeyCommand));
     }
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/build/module/StartupGlobalKeyCommand.java:192]: We probably don't need all the comments here.

File: vassal-app/src/main/java/VASSAL/configure/VisibilityCondition.java (hunk 1 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/configure/VisibilityCondition.java]: fixed

File: vassal-app/src/main/java/VASSAL/build/module/StartupGlobalKeyCommand.java (hunk 2 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/build/module/StartupGlobalKeyCommand.java:181]: lol clean up the body bags

**Pull request merged**